lib/middleware/authentication.js

```diff
@@ -464,35 +497,105 @@ function parseQueryV4(query) {
- *
- * (currently only generates strings for V2 signing)
+ * Generates a string to be signed for signature version 2.
  *
  * @param {*} canonicalRequest
  */
-function getStringToSign(canonicalRequest) {
+function getStringToSignV2(canonicalRequest) {
+  const queryString = canonicalRequest.querystring.replace(/=(&|$)/g, ''); // remove trailing = for empty params
+
   return [
     canonicalRequest.method,
     canonicalRequest.contentMD5,
     canonicalRequest.contentType,
     canonicalRequest.timestamp,
     ...canonicalRequest.amzHeaders,
-    canonicalRequest.querystring
-      ? `${canonicalRequest.uri}?${canonicalRequest.querystring}`
+    queryString
+      ? `${canonicalRequest.uri}?${queryString}`
       : canonicalRequest.uri,
   ].join('\n');
 }
 
+/**
+ * Generates a string to be signed for signature version 4.
+ *
+ * @param {*} canonicalRequest
+ * @param {*} signature
+ */
+function getStringToSignV4(canonicalRequest, { credential, signedHeaders }) {
+  const canonicalHeaders = signedHeaders
+    .map(header => `${header}:${canonicalRequest.headers[header]}\n`)
+    .join('');
+
+  const contentHash =
+    canonicalRequest.headers['x-amz-content-sha256'] || 'UNSIGNED-PAYLOAD';
+
+  const canonicalRequestString = [
+    canonicalRequest.method,
+    canonicalRequest.uri,
+    canonicalRequest.querystring,
+    canonicalHeaders,
+    signedHeaders.join(';'),
+    contentHash,
+  ].join('\n');
+
+  return [
+    'AWS4-HMAC-SHA256',
+    canonicalRequest.timestamp,
+    [
+      credential.date,
+      credential.region,
+      credential.service,
+      credential.termination,
+    ].join('/'),
+    createHash('sha256')
+      .update(canonicalRequestString)
+      .digest('hex'),
+  ].join('\n');
+}
+
 /**
  * Performs the calculation of an authentication code for a string using the specified key and
  * algorithm.
  *
  * @param {String} stringToSign the string representation of a canonical request
- * @param {String} signingKey a secret access key for V2 signing, or a signing key for V4 signing
+ * @param {String} signingKey a secret access key
  * @param {String} algorithm should be one of "sha1" or "sha256"
  */
-function calculateSignature(stringToSign, signingKey, algorithm) {
+function calculateSignatureV2(stringToSign, signingKey, algorithm) {
   const signature = createHmac(algorithm, signingKey);
   signature.update(stringToSign, 'utf8');
   return signature.digest('base64');
 }
+
+/**
+ * Performs the calculation of an authentication code for a string using the specified key and
+ * various components required by v4.
+ *
+ * @param {String} stringToSign the string representation of a canonical request
+ * @param {String} secretKey a secret access key
+ * @param {String} date received from the credential header
+ * @param {String} region received From the credential header
+ * @param {String} service received From the credential header
+ */
+function calculateSignatureV4(stringToSign, secretKey, date, region, service) {
```

Comment:
Do we want to break out the V2 and V4 methods into 2 different files? That way it reduces the file size of this file, and we can separate out concerns a little bit?